pkg/deviceplugins/device_manager.go

```diff
@@ -0,0 +1,490 @@
+	}
+
+	dpi.server = grpc.NewServer([]grpc.ServerOption{}...)
+	defer dpi.stopDevicePlugin()
+
+	pluginapi.RegisterDevicePluginServer(dpi.server, dpi)
+
+	errChan := make(chan error, 2)
+
+	go func() {
+		errChan <- dpi.server.Serve(sock)
+	}()
+
+	err = waitForGRPCServer(dpi.socketPath, connectionTimeout)
+	if err != nil {
+		return fmt.Errorf("error starting the GRPC server: %v", err)
+	}
+
+	err = dpi.register()
+	if err != nil {
+		return fmt.Errorf("error registering with device plugin manager: %v", err)
+	}
+
+	go func() {
+		errChan <- dpi.healthCheck()
+	}()
+
+	dpi.setInitialized(true)
+	logger.Infof("Initialized DevicePlugin: %s", dpi.resourceName)
+	dpi.starter.started = true
+	err = <-errChan
+
+	return err
+}
+
+func (dpi *PCIDevicePlugin) ListAndWatch(_ *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
+	emptyList := []*pluginapi.Device{}
+	s.Send(&pluginapi.ListAndWatchResponse{Devices: dpi.devs})
+
+	done := false
+	for {
+		select {
+		case devHealth := <-dpi.health:
+			for _, dev := range dpi.devs {
+				if devHealth.DevId == dev.ID {
+					dev.Health = devHealth.Health
+				}
+			}
+			s.Send(&pluginapi.ListAndWatchResponse{Devices: dpi.devs})
+			logrus.Infof("Sending ListAndWatchResponse for device with dpi.devs = %v", dpi.devs)
+		case <-dpi.stop:
+			done = true
+		case <-dpi.done:
+			done = true
+		}
+		if done {
+			break
+		}
+	}
+	// Send empty list to increase the chance that the kubelet acts fast on stopped device plugins
+	// There exists no explicit way to deregister devices
+	if err := s.Send(&pluginapi.ListAndWatchResponse{Devices: emptyList}); err != nil {
+		log.DefaultLogger().Reason(err).Infof("%s device plugin failed to deregister: %s", dpi.resourceName, err)
+	}
+	close(dpi.deregistered)
+	return nil
+}
+
+func (dpi *PCIDevicePlugin) Allocate(_ context.Context, r *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	logrus.Infof("Allocate called with %d requests", len(r.ContainerRequests))
+	resourceNameEnvVar := util.ResourceNameToEnvVar(PCIResourcePrefix, dpi.resourceName)
+	allocatedDevices := []string{}
+	resp := new(pluginapi.AllocateResponse)
+	containerResponse := new(pluginapi.ContainerAllocateResponse)
+
+	for _, request := range r.ContainerRequests {
+		deviceSpecs := make([]*pluginapi.DeviceSpec, 0)
+		for _, devID := range request.DevicesIDs {
+			// translate device's iommu group to its pci address
+			devPCIAddress, exist := dpi.iommuToPCIMap[devID]
+			if !exist {
+				continue
+			}
+			allocatedDevices = append(allocatedDevices, devPCIAddress)
+			deviceSpecs = append(deviceSpecs, formatVFIODeviceSpecs(devID)...)
+		}
+		containerResponse.Devices = deviceSpecs
+		envVar := make(map[string]string)
+		envVar[resourceNameEnvVar] = strings.Join(allocatedDevices, ",")
+
+		containerResponse.Envs = envVar
+		resp.ContainerResponses = append(resp.ContainerResponses, containerResponse)
+	}
+	return resp, nil
+}
```

Comment:
```suggestion
func (dpi *PCIDevicePlugin) Allocate(_ context.Context, r *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
	logrus.Debugf("Allocate request %s", r.String())
	resourceNameEnvVar := util.ResourceNameToEnvVar(PCIResourcePrefix, dpi.resourceName)
	allocatedDevices := []string{}
	resp := new(pluginapi.AllocateResponse)
	containerResponse := new(pluginapi.ContainerAllocateResponse)

	for _, request := range r.ContainerRequests {
		deviceSpecs := make([]*pluginapi.DeviceSpec, 0)
		for _, devID := range request.DevicesIDs {
			// translate device's iommu group to its pci address
			logrus.Debugf("looking up deviceID %s in map %v", devID, dpi.iommuToPCIMap)
			iommuGroup, exist := dpi.iommuToPCIMap[devID] // not finding device ids
			if exist {
				// if device exists, check if there other devices
				// in the same iommuGroup, and append these too
				allocatedDevices = append(allocatedDevices, devID)
				for devPCIAddress, ig := range dpi.iommuToPCIMap {
					if ig == iommuGroup {
						allocatedDevices = append(allocatedDevices, devPCIAddress)
					}
				}
				deviceSpecs = append(deviceSpecs, formatVFIODeviceSpecs(iommuGroup)...)
			} else {
				continue // break execution of loop as we are not handling this device
			}
		}
		containerResponse.Devices = deviceSpecs
		envVar := make(map[string]string)
		envVar[resourceNameEnvVar] = strings.Join(allocatedDevices, ",")

		containerResponse.Envs = envVar
		resp.ContainerResponses = append(resp.ContainerResponses, containerResponse)
	}
	logrus.Debugf("Allocate response %v", resp)
	return resp, nil
}
```
Made some changes to Allocate to return correct iommuGroup, and also added additional devices part of the same iommuGroup.
This addresses the error where kubevirt cant create host-devices